File: ast/statement/AssignmentStatement.ts
import { SourceLocation } from "../ASTElement";
import { ExpressionElement } from "../ExpressionElement";
import { StatementElement } from "../StatementElement";


export class AssignmentStatement<LHS extends ExpressionElement, RHS extends ExpressionElement> extends StatementElement {
    lhs: LHS;
    rhs: RHS;

    constructor(loc: SourceLocation, lhs: LHS, rhs: RHS) {
        super(loc);
        this.lhs = lhs;
        this.rhs = rhs;
    }

    clone() {
        return new AssignmentStatement(this.source_location,
            this.lhs.clone() as ExpressionElement,
            this.rhs.clone() as ExpressionElement);
    }

    toString() {
        return `${this.lhs} = ${this.rhs};`;
    }
}

File: backend/src/main/java/dev/m0rg/howl/ast/AssignmentStatement.java
package dev.m0rg.howl.ast;

public class AssignmentStatement extends Statement {
    Expression lhs;
    Expression rhs;

    public AssignmentStatement(Span span) {
        super(span);
    }

    @Override
    public String format() {
        return this.lhs.format() + " = " + this.rhs.format() + ";";
    }

    @Override
    public ASTElement detach() {
        AssignmentStatement rc = new AssignmentStatement(span);
        rc.setLHS((Expression) this.lhs.detach());
        rc.setRHS((Expression) this.rhs.detach());
        return rc;
    }

    public void setLHS(Expression lhs) {
        this.lhs = (Expression) lhs.setParent(this);
    }

    public void setRHS(Expression rhs) {
        this.rhs = (Expression) rhs.setParent(this);
    }

    public void transform(ASTTransformer t) {
        this.lhs.transform(t);
        this.setLHS(t.transform(lhs));
        this.rhs.transform(t);
        this.setRHS(t.transform(rhs));
    }
}
